### PR TITLE
ENH Remove blank strings

### PR DESCRIPTION
**Pull request by emteknetnz**
Issue https://github.com/silverstripe/.github/issues/14

### CODE REVIEW

File: src/Translator.php [hunk 1]
```diff
@@ -262,11 +263,15 @@ private function transifexPullSource()
...
         $this->log('Merging local yaml files');
+        // $this->originalYaml at this point is yaml prior to adding in transifex yaml
         foreach ($this->originalYaml as $path => $contentYaml) {
             // If there are any keys in the original yaml that are missing now, add them back in.
             if (file_exists($path)) {
                 $rawYaml = file_get_contents($path);
+                // $parsedYaml at this point is transifex yaml
                 $parsedYaml = Yaml::parse($rawYaml);
+                $this->removeBlankStrings($parsedYaml);
+                // transifex yaml has precidence over original yaml
```
Comment (michalkleiner) [inline src/Translator.php]:
```suggestion
                // transifex yaml has precedence over original yaml
```

File: src/Translator.php [hunk 2]
```diff
@@ -206,7 +206,8 @@ private function storeYaml(): void
         foreach ($this->modulePaths as $modulePath) {
             foreach (glob($this->getYmlLangDirectory($modulePath) . '/*.yml') as $path) {
                 $rawYaml = file_get_contents($path);
-                $this->originalYaml[$path] = Yaml::parse($rawYaml);
+                $yaml = Yaml::parse($rawYaml);
+                $this->originalYaml[$path] = $yaml;
```
Comment (michalkleiner) [inline src/Translator.php]: Unnecessary change left over from debugging?

Comment (emteknetnz) [reply src/Translator.php]: Updated

File: src/Translator.php (hunk 1 above)
Comment (emteknetnz) [reply src/Translator.php]: Updated

**Review by GuySartorelli — COMMENTED** [review]
Should we validate that the blank strings aren't _also_ present in the original english source before removing them?

Also, given that we use transifex for json and yaml in near identical ways, should we defensively add this to the json handling as well?

**Comment by emteknetnz** [comment]
I've updated the PR to filter out strings where the en.yml version is blank, though I'd be surprised if there are any source translations with a blank string.  I've tried rerunning tx-translator with this update and thing still appear to work correctly

I couldn't find any blank strings in json, so we don't really have a need right now for this, nor a good way to validate it.  If we see blank strings in the future I think that's the time to add in support for it

File: src/Translator.php [hunk 3]
```diff
@@ -275,6 +279,51 @@ private function mergeYaml(): void
...
+        }
+    }
+
+    private function recursiveRemoveBlankStrings(string $parentKey, array &$data)
+    {
+        foreach ($data as $key => &$val) {
+            if (is_array($val)) {
+                $this->recursiveRemoveBlankStrings($parentKey . '-' . $key, $val);
+            }
+            if (empty($val) && !array_key_exists($parentKey . '-' . $key, $this->blankEnStrings)) {
```
Comment (ChloeSartorelli) [inline src/Translator.php]:
```suggestion
            if ($val === '' && !array_key_exists($parentKey . '-' . $key, $this->blankEnStrings)) {
```
Definitely an edge case, but protecting against a case where the translation string is `'0'` which PHP considers empty.

**Review by GuySartorelli — CHANGES_REQUESTED** [review]

File: src/Translator.php (hunk 3 above)
Comment (emteknetnz) [reply src/Translator.php]: Updated.

**Review by GuySartorelli — APPROVED** [review]
Looks good, thanks

**Comment by GuySartorelli** [comment]
Tagged as [1.0.3](https://github.com/silverstripe/silverstripe-tx-translator/releases/tag/1.0.3)
